Lib9c/Action/UnlockEquipmentRecipe.cs

```diff
@@ -0,0 +1,123 @@
+    public class UnlockEquipmentRecipe: GameAction
+    {
+        public List<int> RecipeIds = new List<int>();
+        public Address AvatarAddress;
+
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            var states = context.PreviousStates;
+            var worldInformationAddress = AvatarAddress.Derive(LegacyWorldInformationKey);
+            var questListAddress = AvatarAddress.Derive(LegacyQuestListKey);
+            var inventoryAddress = AvatarAddress.Derive(LegacyInventoryKey);
+            var unlockedRecipeIdsAddress = AvatarAddress.Derive("recipe_ids");
+            if (context.Rehearsal)
+            {
+                return states
+                    .SetState(worldInformationAddress, MarkChanged)
+                    .SetState(questListAddress, MarkChanged)
+                    .SetState(inventoryAddress, MarkChanged)
+                    .SetState(AvatarAddress, MarkChanged)
+                    .SetState(unlockedRecipeIdsAddress, MarkChanged)
+                    .MarkBalanceChanged(GoldCurrencyMock, AvatarAddress);
+            }
+
+            WorldInformation worldInformation;
+            bool migrationRequired = false;
+            AvatarState avatarState = null;
+            if (states.TryGetState(worldInformationAddress, out Dictionary rawInfo))
+            {
+                worldInformation = new WorldInformation(rawInfo);
+            }
+            else
+            {
+                // AvatarState migration required.
+                if (states.TryGetAvatarState(context.Signer, AvatarAddress, out avatarState))
+                {
+                    worldInformation = avatarState.worldInformation;
+                    migrationRequired = true;
+                }
+                else
+                {
+                    // Invalid Address.
+                    throw new FailedLoadStateException($"Can't find AvatarState {AvatarAddress}");
+                }
+            }
+
+            var equipmentRecipeSheet = states.GetSheet<EquipmentItemRecipeSheet>();
+
+            List<int> unlockedIds = states.TryGetState(unlockedRecipeIdsAddress, out List rawIds)
+                ? rawIds.ToList(StateExtensions.ToInteger)
+                : new List<int>();
+
+            int totalCost = 0;
+            foreach (var recipeId in RecipeIds)
+            {
+                if (unlockedIds.Contains(recipeId))
+                {
+                    // Already Unlocked
+                    throw new AlreadyRecipeUnlockedException($"recipe: {recipeId} already unlocked.");
+                }
+                unlockedIds.Add(recipeId);
+
+                EquipmentItemRecipeSheet.Row row = equipmentRecipeSheet[recipeId];
+                if (!worldInformation.IsStageCleared(row.UnlockStage))
+                {
+                    throw new NotEnoughClearedStageLevelException($"clear {row.UnlockStage} first.");
+                }
+
+                totalCost += 500;
+            }
```

Comment:
I think you should also throw exception when recipe which has lower `recipeId` exists.